File: 0045-jump-game-ii/0045-jump-game-ii.cpp
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, INT_MAX);
        dp[n-1] = 0;
        for(int i=n-2; i>=0; i--) {
            int steps = min(n-1, i + nums[i]);
            for(int j=i+1; j<=steps; j++) {
                if(dp[j] != INT_MAX) {
                    dp[i] = min(dp[i], dp[j] + 1);
                }
            }
        }
        return dp[0];
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.rs
impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut dp = vec![std::i32::MAX; n];
        dp[n-1] = 0;
        for i in (0..n-1).rev() {
            let steps = std::cmp::min(n-1, i + nums[i] as usize);
            for j in (i+1..=steps) {
                if dp[j] != std::i32::MAX {
                    dp[i] = std::cmp::min(dp[i], dp[j] + 1);
                }
            }
        }
        dp[0]
    }
}
